### ISSUE TITLE
grow_tree and tree matrix

### ISSUE THREAD

**Issue by EoghanONeill** [BelindaHernandez/bartBMA#6]
I'm not sure if I'm interpreting the tree matrices correctly. I assume that column numbers correspond to the depth. 

Suppose that some terminal nodes are up to depth 4 (5). If there is also a terminal node numbered 2 (for example) of depth 1 (2), then I am assuming that the rows corresponding to observations that fall in node 2 should equal one for the zeroth (first) column, 2 for the first (second) column, and also 2 for the second (third) to fourth (fifth) columns.
Under this assumption, no elements of the tree matrix should be equal to zero.

Assuming that the above assumptions are correct, there are the following potential issues in the grow_tree function:

Recall the example in issue number 4 (R code below).

```
library(devtools)
install_github("EoghanONeill/bartBMA")
library(bartBMA)

#simulate data
N <- 1000
set.seed(100)
x1 <- runif(N)
x2 <- runif(N)
x3 <- runif(N)
x4 <- runif(N)
x5 <- runif(N)
x6 <- runif(N)
x7 <- runif(N)
x8 <- runif(N)
x9 <- runif(N)
x10 <- runif(N)
epsilon <- rnorm(N)

xcov <- cbind(x1,x2,x3,x4,x5,x6,x7,x8,x9,x10)
y <- sin(pi*x1*x2) + 20*(x3-0.5)^2+10*x4+5*x5+epsilon

bart_bma_example <- bartBMA(x.train = xcov, y.train = y)


tree_mat_issue <- bart_bma_example[[3]][[1]][[1]] 
```

We can see why the new tree matrix can have a final column with the final two node numbers in the rightmost columns, and zero otherwise.
Line 487 of BARTBMA_SumTreeLikelihood.cpp uses the function addcol, but while this correctly numbers the observations sent left and right when the node is grown (i.e. when the column is added), it leaves all other rows as zero (for the rightmost column).

Perhaps there should be another few lines after line 487 that loops over all rows of prior_tree_matrix_temp, and then contains an if statement (within the loop) to ensure that the row is not in grow_node, and then (within the if statement) set the final column entry for the row equal to the second last column entry for that row.

I include some untested C++ code below to make this idea clear:
```
for(int i=0;i<prior_tree_matrix_temp.nrow();i++){
 if(std::find(grow_node(), grow_node(), i) == grow_node() ){
   prior_tree_matrix_temp(i,prior_tree_matrix_temp.ncol()-1) = prior_tree_matrix_temp(i,prior_tree_matrix_temp.ncol()-2);
 } 
}
```
 However, perhaps another issue is: Is the last row of the tree table always a node of the greatest depth reached so far? i.e. is it true that none of the earlier rows of the tree table refer to greater depth nodes?

.

.

For splitting nodes that are not on the final row of the tree table, there may be another issue with the tree depth input used for grow_tree in get_best_split and get_best_split_sum (lines 954, 1138)

I include some R code below:
```
# Consider the following two lines, 928, 930, in the C++ code.

find_term_cols(tree_mat_issue,11) # NOTE: column index output starts at 0 for first column
find_term_cols(tree_mat_issue,1)
find_term_cols(tree_mat_issue,3)

sourceCpp( code = '
// [[Rcpp::depends("RcppArmadillo")]]
           
           #include <RcppArmadillo.h>
           
           using namespace arma ; 
           using namespace Rcpp ;
           
           // [[Rcpp::export]]
           NumericVector unique_example( NumericVector vec_example){
           NumericVector X1 = unique(vec_example); 
           return X1;
           }
           ')

unique(find_term_cols(tree_mat_issue,3)) # R function does not sort descending
unique_example(find_term_cols(tree_mat_issue,3)) # C++ function sorts descending
#indexes the columns starting with 0.

#The lines in the C++ code are 928 and 929 (also 918, 920, 1102, 1104, 1114):
#d1=unique(find_term_cols(treemat_c,terminal_nodes[l]));
#d=d1[0];

# unique sorts DESCENDING, and therefore d1[0] gives 
# the index of the rightmost column with at least one element equal to terminal_nodes[l].
# But the leftmost would represent the true depth (or depth plus one depending on indexing)
# of the terminal node

#I assume that d should represent the depth?
#If d is incorrect, then grow_tree will update the tree matrix
# by changing the node number from terminal_node to the daughter node numbers
# at the incorrect depth (i.e. in the incorrect column).

# Perhaps the following functions would also have to be edited if 
#lines 928 and 929 (also 918, 920, 1102, 1104, 1114) were edited:
#update_grow_obs
#set_daughter_to_end_mat
#find_obs_to_update_grow

# but I don't know what needs to be edited.

```

**Comment by EoghanONeill** [comment]
This issue is probably solved by the solution to issue number  #4 

My first point about the tree matrix is irrelevant because I didn't know that a tree matrix row should contain zeros past the final depth that an observation reaches.

My second point about using the index of the rightmost or leftmost column is now incorrect because the rightmost column should give the correct depth now that issue  #4 is fixed.

**Comment by BelindaHernandez** [comment]
agree this one is solved by the fix for #4 so am going to close
